src/portal/mod.rs

```diff
@@ -218,6 +238,48 @@ impl Keyring {
     }
 }
 
+pub trait KeyExt {
+    fn get<'a>(&'a self) -> Pin<Box<dyn std::future::Future<Output = &crate::Key> + 'a>>;
+}
+
+impl KeyExt for Keyring {
+    fn get<'a>(&'a self) -> Pin<Box<dyn std::future::Future<Output = &crate::Key> + 'a>> {
+        Box::pin(async {
+            if let Some(key) = self.key.get() {
+                key
+            } else {
+                let keyring = self.keyring.lock().await;
+                self.key.get_or_init(|| keyring.derive_key(&self.secret))
+            }
```

Comment:
Can't this be a self.key.get_or_init directly without the whole condition?

Reply:
I see. That is not great but it is fine, could use a comment why it is done this to avoid future you or me trying to change it :P